### ISSUE TITLE
Update github.com/coreos/go-oidc package to latest

### ISSUE THREAD

**Issue by ericchiang** [kubernetes/kubernetes#42654]
A few months ago we update go-oidc after having a lot of issues with compliance around JWT parsing, the OAuth2 spec, cache header parsing, etc. The new package `github.com/coreos/go-oidc` replaces the existing packages `github.com/coreos/go-oidc/{http,jose,keys,oauth2,oidc}`.

Over the 1.7 release cycle we'd like to update the packages use for the API server and kubectl plugins. As part of this work, we'd like to also introduce some e2e tests for the OIDC plugin, so we can continue to ensure it functions properly.

Based on my estimations, this update would fix:

* https://github.com/kubernetes/kubernetes/issues/37875 (oauth2 compliance)
* https://github.com/coreos/go-oidc/issues/136 (cache header parsing)
* https://github.com/coreos/go-oidc/issues/137 (less picky about metadata claims)

cc @mlbiam @frodenas @curtisallen @jsloyer @rithujohn191 @kubernetes/sig-auth-misc

**Comment by curtisallen** [comment]
Yeah, I agree e2e tests would be very beneficial
👍

**Comment by ericchiang** [comment]
@curtisallen the fix for coreos/go-oidc#137 is a bug fix on our end. I'd advocating sneaking that patch in for 1.6. I don't see us fixing the other issues before 1.7.

**Comment by ericchiang** [comment]
Also when we do this I'd like to take a look at modifying the claim prefixing behavior (https://github.com/kubernetes/kubernetes/issues/31380). Maybe let users just prefix values with `oidc:` instead of the full issuer URL.

**Comment by frodenas** [comment]
@ericchiang Is there any progress on this issue? Do you believe the updated packages will land in 1.7? Thanks!

**Comment by ericchiang** [comment]
@frodenas yes this is still slated for 1.7. I'll update with the relevant PRs when I get around to drafting them up.

**Comment by frodenas** [comment]
Thanks for the heads up @ericchiang!

**Comment by philips** [comment]
One side effect that I hope this resolves is the needless roundtrips to the identity provider:

```
$ kubectl get pods  -v=10 2>&1 | grep identity
I0418 20:26:45.683891   26772 round_trippers.go:299] curl -k -v -XGET  https://tectonic.example.com/identity/.well-known/openid-configuration
I0418 20:26:45.876712   26772 round_trippers.go:318] GET https://tectonic.example.com/identity/.well-known/openid-configuration 200 OK in 192 milliseconds
```

Ideally these only happen when a tokens request is required.

**Comment by thomastaylor312** [comment]
This didn't land in 1.7, is there anything holding this up or could I submit a patch for this? The expires header parsing issue is causing problems for us trying to use Okta

**Comment by ericchiang** [comment]
@thomastaylor312 can you send a fix for https://github.com/coreos/go-oidc/issues/136 ? I was hoping to move away from those subpackages since we're trying to move to the new package, but updating the old one and doing a small bump here seems reasonable.

**Comment by thomastaylor312** [comment]
Yeah, I'll take a look at it. Thanks @ericchiang

**Comment by thomastaylor312** [comment]
PR open to fix the header parsing: https://github.com/coreos/go-oidc/pull/153

**Comment by thomastaylor312** [comment]
@ericchiang The PR with the updated Godeps for the new go-oidc code is open here: #48907

**Comment by wardviaene** [comment]
I'm hitting a bug present in the go-oidc version kubernetes is using.
https://github.com/coreos/go-oidc/issues/66 fixes a trailing slash issue if you're using Auth0

**Comment by liggitt** [comment]
kubernetes already has picked up that fix and has it available in 1.7.x:

https://github.com/kubernetes/kubernetes/blob/701a4e3a7c115e5b32515bffcef16b94177793c1/vendor/github.com/coreos/go-oidc/oidc/provider.go#L621-L624

**Comment by ericchiang** [comment]
Going to close this issue.

Having shipped the old version of the oidc package in several releases, and hit lots of caveats with lots of providers, I don't think it's worth the risk of updating to the new version of the package. If there's an explicit reason we can do it in the future, but for now we'll continue to maintain the old version for Kubernetes' use.

/close
